### PR TITLE
WIP: use netcdf/cf inspired model for tracking CRS information

### LINKED ISSUE #837: Understand NetCDF CF convention for projection info in `.geobox` property
### Expected behaviour

`xr.open_dataset(<netcdf CF file>).geobox is not None`

### Actual behaviour

netcdf cf files loaded via `xr.open_dataset` are seen as not-georegistered by datcube `.geobox` extension. Even for files produced by datacube itself. 

Also see comment thread in https://github.com/opendatacube/datacube-core/pull/835#issuecomment-564590677

### PR DESCRIPTION
**Pull request by Kirill888**
### Reason for this pull request

To address issue #837. Rather than rely on xarray attributes that are easily lost use non-dimensional coordinate to capture CRS WKT information.

### Proposed changes

1. Update on-disk representation of ingester generated files
   - mark grid_mapping variable as a coordinate
   - rename grid_mapping variable `crs->spatial_ref` to not clash with crs attribute
2. Update `dc.load` to create xarrays with appropriate structure
3. Update `.geobox` code to understand data from 2,3

 - [x] Closes #837 
 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: datacube/api/core.py [hunk 1]
```diff
@@ -427,38 +425,46 @@ def create_storage(coords, geobox, measurements, data_func=None):
         :param measurements:
             list of :class:`datacube.model.Measurement`
 
-        :param data_func:
+        :param data_func: Callable `Measurement -> np.ndarray`
             function to fill the storage with data. It is called once for each measurement, with the measurement
-            as an argument. It should return an appropriately shaped numpy array. If not provided, an empty
-            :class:`xarray.Dataset` is returned.
+            as an argument. It should return an appropriately shaped numpy array. If not provided memory is
+            allocated an filed with `nodata` value defined on a given Measurement.
```
Comment (uchchwhash) [inline datacube/api/core.py:432]: an filed -> and filled?

File: datacube/ui/task_app.py [hunk 2]
```diff
@@ -247,12 +241,6 @@ def check_existing_files(paths):
     ))
 
 
-def add_dataset_to_db(index, datasets):
-    for dataset in datasets.values:
-        index.datasets.add(dataset, with_lineage=False)
-        _LOG.info('Dataset added')
-
-
 def do_nothing(result):
```
Comment (uchchwhash) [inline datacube/ui/task_app.py:244]: What is this about?

**Review by uchchwhash — APPROVED** [review]
I am OK with this. :+1:

File: datacube/ui/task_app.py (hunk 2 above)
Comment (Kirill888) [reply datacube/ui/task_app.py:244]: Part of this commit https://github.com/opendatacube/datacube-core/pull/899/commits/c6c4167283c74c46b3eaa67e5a83b361afa3339b just house-keeping, removing dead code that is not covered by tests.

File: datacube/api/core.py (hunk 1 above)
Comment (Kirill888) [reply datacube/api/core.py:432]: thanks, fixed

File: datacube/ui/task_app.py (hunk 2 above)
Comment (uchchwhash) [reply datacube/ui/task_app.py:244]: cool!

**Pull request merged**